python/ccxt/base/exchange.py

```diff
@@ -365,17 +365,16 @@ def fetch(self, url, method='GET', headers=None, body=None):
             self.raise_error(ExchangeError, url, method, e)
 
         except HTTPError as e:
-            self.handle_errors(response.status_code, response.reason, url, method, self.last_response_headers, self.last_http_response)
-            self.handle_rest_errors(e, response.status_code, self.last_http_response, url, method)
-            self.raise_error(ExchangeError, url, method, e, self.last_http_response)
+            self.handle_errors(response.status_code, response.reason, url, method, response.headers, response.text)
```

Comment:
This attempt I like much better, however... 

The initial problem/goal is to make this particular call to handle_errors accept a JSON-decoded response, if any,(under a condition to parse the JSON just once for the entire response). And this call hasn't changed ) The rest of the changes, however, a less significant, meaning that we apart from the added argument in [line R408](https://github.com/ccxt/ccxt/pull/2296/files#diff-4ca611a196a4dc7a1e1f1170b47c5537R408), other changes are mostly renaming and restructuring. However, we're not approaching the goal yet )

So, returning to the initial post:

1. Make the HTTP request
2. Parse JSON response once, if any, whatever the http response code
3. Pass parsed JSON response to handleErrors along with other data needed for exchange-specific error handling, including the http response code, whatever it is
4. In `handleErrors` handle any exchange-specific errors signalled either in a JSON-encoded format or with HTTP codes
5. In `handleRestErrors` handle HTTP error codes that were not intercepted in handleErrors and intercept common DDoSProtection pages...
 
It now does not really implement p1+p2+p3 in case of a HTTP error... The task itself may be really a little more involved than it seems at first )

Reply:
Perhaps best way to attempt this would be to use a `try: except` and attempt to set the "maybe json" as an attribute of the response object. If the parsing failes we could set the json attribute to None and check it later in handleRestResponse.

Reply:
There may be a simpler solution to this, I mean, we could add a call to `parse_json` to the `except HTTPError` case and from there pass the resulting JSON (if any) to the `handle_errors`. Wouldn't it do the job?

Reply:
Could the two lines underneath except HTTPError be replaced with a single `self.handRestResponse` and then a `self.raise_error`?

Reply:
@frosty00 yep, I think it should do what we want, exactly